Mathlib/CategoryTheory/Adjunction/PartialAdjoint.lean

```diff
@@ -0,0 +1,166 @@
+noncomputable def partialLeftAdjointObj (X : F.PartialLeftAdjointSource) : D :=
+  (F ⋙ coyoneda.obj (op X.obj)).coreprX
+
+/-- Given `F : D ⥤ C`, this is the canonical bijection
+`(F.partialLeftAdjointObj X ⟶ Y) ≃ (X.obj ⟶ F.obj Y)`
+for all `X : F.PartialLeftAdjointSource` and `Y : D`. -/
+noncomputable def partialLeftAdjointHomEquiv {X : F.PartialLeftAdjointSource} {Y : D} :
+    (F.partialLeftAdjointObj X ⟶ Y) ≃ (X.obj ⟶ F.obj Y) :=
+  (F ⋙ coyoneda.obj (op X.obj)).corepresentableBy.homEquiv
+
+lemma partialLeftAdjointHomEquiv_comp {X : F.PartialLeftAdjointSource} {Y Y' : D}
+    (f : F.partialLeftAdjointObj X ⟶ Y) (g : Y ⟶ Y') :
+    F.partialLeftAdjointHomEquiv (f ≫ g) =
+      F.partialLeftAdjointHomEquiv f ≫ F.map g := by
+  apply CorepresentableBy.homEquiv_comp
+
+/-- Given `F : D ⥤ C`, this is `F.partialLeftAdjoint` on morphisms. -/
+noncomputable def partialLeftAdjointMap {X Y : F.PartialLeftAdjointSource}
+    (f : X ⟶ Y) : F.partialLeftAdjointObj X ⟶ F.partialLeftAdjointObj Y :=
+    F.partialLeftAdjointHomEquiv.symm (f ≫ F.partialLeftAdjointHomEquiv (𝟙 _))
+
+@[simp]
+lemma partialLeftAdjointHomEquiv_map {X Y : F.PartialLeftAdjointSource}
+    (f : X ⟶ Y) :
+    F.partialLeftAdjointHomEquiv (F.partialLeftAdjointMap f) =
+      by exact f ≫ F.partialLeftAdjointHomEquiv (𝟙 _) := by
+  simp [partialLeftAdjointMap]
+
+lemma partialLeftAdjointHomEquiv_map_comp {X X' : F.PartialLeftAdjointSource} {Y : D}
+    (f : X ⟶ X') (g : F.partialLeftAdjointObj X' ⟶ Y) :
+    F.partialLeftAdjointHomEquiv (F.partialLeftAdjointMap f ≫ g) =
+      by exact f ≫ F.partialLeftAdjointHomEquiv g := by
+  rw [partialLeftAdjointHomEquiv_comp, partialLeftAdjointHomEquiv_map, assoc,
+    ← partialLeftAdjointHomEquiv_comp, id_comp]
+
+/-- Given `F : D ⥤ C`, this is the partial adjoint functor `F.PartialLeftAdjointSource ⥤ D`. -/
+@[simps]
+noncomputable def partialLeftAdjoint : F.PartialLeftAdjointSource ⥤ D where
+  obj := F.partialLeftAdjointObj
+  map := F.partialLeftAdjointMap
+  map_id X := by
+    apply F.partialLeftAdjointHomEquiv.injective
+    dsimp
+    rw [partialLeftAdjointHomEquiv_map]
+    erw [id_comp]
+  map_comp {X Y Z} f g := by
+    apply F.partialLeftAdjointHomEquiv.injective
+    dsimp
+    rw [partialLeftAdjointHomEquiv_map, partialLeftAdjointHomEquiv_comp,
+      partialLeftAdjointHomEquiv_map, assoc]
+    erw [assoc]
+    rw [← F.partialLeftAdjointHomEquiv_comp, id_comp,
+      partialLeftAdjointHomEquiv_map]
+
+variable {F}
+
+lemma isRightAdjoint_of_leftAdjointObjIsDefined_eq_top
+    (h : F.LeftAdjointObjIsDefined = ⊤) : F.IsRightAdjoint := by
+  replace h : ∀ X, IsCorepresentable (F ⋙ coyoneda.obj (op X)) := fun X ↦ by
+    simp only [← leftAdjointObjIsDefined_iff, h, Pi.top_apply, Prop.top_eq_true]
+  exact (Adjunction.adjunctionOfEquivLeft
+    (fun X Y ↦ (F ⋙ coyoneda.obj (op X)).corepresentableBy.homEquiv)
+    (fun X Y Y' g f ↦ by apply CorepresentableBy.homEquiv_comp)).isRightAdjoint
```

Comment:
Would be nice to have the converse if it isn't too hard.

Reply:
Thanks for the suggestion, I have added the iff lemma in https://github.com/leanprover-community/mathlib4/pull/17388/commits/7ca7077d174a78fc13926b8aa9799f7b134a97ca